File: futpoolsapp/futpoolsapp/Models/Quiniela.swift
//
//  Quiniela.swift
//  futpoolsapp
//

import Foundation

struct Quiniela: Decodable, Identifiable {
    let id: String
    let name: String
    let description: String?
    let prize: String
    let cost: String
    let currency: String?
    let startDate: String
    let endDate: String?
    let fixtures: [QuinielaFixture]
    let entriesCount: Int?
    /// Pool lifecycle: `scheduled` (not started), `live` (in progress), `completed` (finished).
    let status: String?
    /// Admin-controlled "featured" flag. When true the pool is pinned to the Home
    /// QUICK PLAY hero (or the featured carousel when multiple are marked).
    let featured: Bool?
    /// User id of the creator (null for legacy admin-seeded pools).
    let createdBy: String?
    /// Resolved username of the creator (from backend populate), used for "Created by @x" badge.
    let createdByUsername: String?
    let createdByDisplayName: String?
    /// `public` → discoverable in Home. `private` → only reachable via inviteCode.
    let visibility: String?
    /// Short alphanum code used in share links (e.g. `futpools://p/ABC23456`).
    let inviteCode: String?
    /// Free-text label for MVP user-pools ("the loser buys pizza"). Not money.
    let prizeLabel: String?
    /// v3 economy — filled by backend. nil on legacy pools.
    let fundingModel: String?
    let platformPrizeCoins: Int?
    let entryCostCoins: Int?
    let rakePercent: Int?
    /// simple_version per-entry MXN fee. Replaces the legacy coin-based
    /// economy for pools created via /admin/pools/new. Optional so the
    /// model still decodes legacy pools that don't carry it.
    let entryFeeMXN: Int?
    /// Real-world prize attached to admin-curated pools. When set,
    /// the detail view renders the prize hero image + AMOE + Apple
    /// Guideline 5.3 disclaimers, and the pool surfaces in the Home
    /// "WEEKLY POOL · REAL PRIZE" teaser.
    let realPrize: RealPrize?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name
        case description
        case prize
        case cost
        case currency
        case startDate
        case endDate
        case fixtures
        case entriesCount
        case status
        case featured
        case createdBy
        case createdByUsername
        case createdByDisplayName
        case visibility
        case inviteCode
        case prizeLabel
        case fundingModel
        case platformPrizeCoins
        case entryCostCoins
        case rakePercent
        case entryFeeMXN
        case realPrize
    }

    var startDateValue: Date? {
        DateParser.parse(startDate)
    }

    var endDateValue: Date? {
        guard let endDate else { return nil }
        return DateParser.parse(endDate)
    }

    var isScheduled: Bool { status == "scheduled" }
    var isLive: Bool { status == "live" }
    var isCompleted: Bool { status == "completed" }

    /// User-facing entry fee. Prefers the new MXN field; falls back to
    /// the legacy coin string for pools created on master. Returns
    /// "—" when neither is set.
    var entryFeeDisplay: String {
        if let mxn = entryFeeMXN, mxn > 0 { return "$\(mxn) MXN" }
        let trimmed = cost.trimmingCharacters(in: .whitespaces)
        return (!trimmed.isEmpty && trimmed != "0") ? cost : "—"
    }

    /// Live-computed prize pool. For simple_version pools (entryFeeMXN
    /// set) the prize is `entries × fee × (1 - rake/100)` — the rake
    /// defaults to 10%. For legacy pools fall back to the prize string
    /// the admin set.
    var prizePoolDisplay: String {
        if let mxn = entryFeeMXN, mxn > 0 {
            let entries = entriesCount ?? 0
            let rake = Double(rakePercent ?? 10) / 100.0
            let pot = Int(Double(entries) * Double(mxn) * (1.0 - rake))
            return pot > 0 ? "$\(pot) MXN" : "—"
        }
        let trimmed = prize.trimmingCharacters(in: .whitespaces)
        return trimmed.isEmpty ? "—" : prize
    }

    /// Backend can wedge a finished pool at `status: "live"` when the
    /// API-Football status fetch fails (no data → fallback marks past
    /// kickoffs as anyStarted but never as allFinished). Override the
    /// label client-side once `endDate` is more than 6h in the past so a
    /// 5-day-old pool stops glowing red on Home.
    var effectiveStatus: String? {
        if status == "live", let end = endDateValue, end < Date().addingTimeInterval(-6 * 3600) {
            return "completed"
        }
        return status
    }

    /// Parsed entry cost for balance checks (e.g. 35 from "35" or "$35").
    var entryCostValue: Double {
        let trimmed = cost.trimmingCharacters(in: .whitespaces)
        let digits = trimmed.filter { $0.isNumber || $0 == "." || $0 == "-" }
        return Double(digits) ?? 0
    }
}

struct QuinielaFixture: Decodable, Identifiable {
    let fixtureId: Int
    let leagueId: Int?
    let leagueName: String?
    let homeTeamId: Int?
    let awayTeamId: Int?
    let homeTeam: String
    let awayTeam: String
    let homeLogo: String?
    let awayLogo: String?
    let kickoff: String
    let status: String?

    var id: Int { fixtureId }

    var kickoffDate: Date? {
        DateParser.parse(kickoff)
    }
}

struct QuinielaUpdateRequest: Encodable {
    let name: String
    let description: String?
    let prize: String
    let cost: String
    let currency: String?
    let featured: Bool?
}

/// Partial update for admin quick-actions (e.g. toggle the featured flag) without
/// having to resend every editable field.
struct QuinielaFeaturedRequest: Encodable {
    let featured: Bool
}

/// Body for `POST /quinielas` — any authenticated user can create a pool.
/// v3: mutually exclusive economy picks.
///   - `entryCostCoins > 0` → peer pool, all participants pay (presets 10/25/50/100/250/500).
///   - `prizeCoins > 0`     → sponsored pool, creator pays prize × 1.1 upfront
///     (presets 50/100/250/500/1000). Participants play free.
/// Server rejects if both are > 0 simultaneously.
struct QuinielaCreateRequest: Encodable {
    let name: String
    let description: String?
    let prizeLabel: String?
    let visibility: String // "public" | "private"
    let entryCostCoins: Int
    let prizeCoins: Int
    let fixtures: [QuinielaCreateFixture]
    /// Optional admin-only real-world prize. Backend silently drops
    /// this field for non-admin tokens, so it's safe to attach
    /// unconditionally from the form.
    let realPrize: RealPrize?
}

/// Real-world prize attached to a pool. Shape mirrors the backend
/// `realPrizeSchema` subdoc on Quiniela. `imageKey` is a bundled
/// asset identifier the iOS app resolves locally
/// (e.g. "PrizeAmazonGift").
struct RealPrize: Codable {
    let label: String
    let prizeUSD: Int
    let imageKey: String?
    let deliveryNote: String?
}

struct QuinielaCreateFixture: Encodable {
    let fixtureId: Int
    let leagueId: Int?
    let leagueName: String?
    let homeTeamId: Int?
    let awayTeamId: Int?
    let homeTeam: String
    let awayTeam: String
    let homeLogo: String?
    let awayLogo: String?
    /// ISO-8601 UTC kickoff.
    let kickoff: String
}

struct QuinielaEntryRequest: Encodable {
    let picks: [QuinielaPick]
}

struct QuinielaPick: Encodable {
    let fixtureId: Int
    let pick: String
}

struct QuinielaEntry: Decodable, Identifiable {
    let id: String
    let quiniela: Quiniela
    let user: String?
    let entryNumber: Int?
    let picks: [QuinielaPickResponse]
    let createdAt: String?
    let updatedAt: String?
    /// Points for this entry (1 per correct forecast). From backend when available.
    let score: Int?
    /// Total fixtures with a final result in this quiniela (max possible points).
    let totalPossible: Int?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case quiniela
        case user
        case entryNumber
        case picks
        case createdAt
        case updatedAt
        case score
        case totalPossible
    }

    var createdAtValue: Date? {
        guard let createdAt else { return nil }
        return DateParser.parse(createdAt)
    }
}

struct QuinielaPickResponse: Decodable {
    let fixtureId: Int
    let pick: String
}

/// Wrapper used when decoding arrays of entries from the backend so a
/// single malformed row (e.g. `quiniela: null` left behind by a deleted
/// pool) doesn't fail the entire response. Decode succeeds for every
/// row; `value` is nil for rows that didn't satisfy `QuinielaEntry`.
struct FailableQuinielaEntry: Decodable {
    let value: QuinielaEntry?
    init(from decoder: Decoder) throws {
        value = try? QuinielaEntry(from: decoder)
    }
}

/// Response shape for `GET /quinielas/:id/participants` — creator-only admin
/// view. Picks are bundled per entry once the pool starts (status !=
/// "scheduled"); during the scheduled phase the backend strips them and
/// sets `picksHidden: true` so the creator can't kick a player based on who
/// guessed well. Web parity: futpools_backend/.../quinielaController.js
/// getParticipants.
struct ParticipantsResponse: Decodable {
    let status: String
    let picksHidden: Bool?
    let participants: [ParticipantDTO]
}

struct ParticipantDTO: Decodable, Identifiable {
    let user: ParticipantUser
    let entryCount: Int
    let firstEntryAt: String?
    let entries: [ParticipantEntry]
    var id: String { user.id }
}

struct ParticipantUser: Decodable {
    let id: String
    let username: String?
    let displayName: String?
}

struct ParticipantEntry: Decodable, Identifiable {
    let _id: String
    let entryNumber: Int
    let createdAt: String?
    /// Optional — present only when the backend reveals picks (post-kickoff).
    /// Nil during scheduled phase.
    let picks: [ParticipantPick]?
    /// Optional scoring stamp. Both nil until first FT in the pool.
    let score: Int?
    let totalPossible: Int?
    var id: String { _id }
}

struct ParticipantPick: Decodable, Hashable {
    let fixtureId: Int
    let pick: String
}

/// Response from `DELETE /quinielas/:id/entries/:entryId`.
struct DeleteEntryResponse: Decodable {
    let ok: Bool
    let refundedAmount: Int?
    let entryId: String?
}

enum DateParser {
    private static let isoWithFractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let isoBasic: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    static func parse(_ value: String) -> Date? {
        if let d = isoWithFractional.date(from: value) { return d }
        if let d = isoBasic.date(from: value) { return d }
        return nil
    }
}

File: futpools_backend/src/models/Quiniela.js
const mongoose = require('mongoose');

const fixtureSchema = new mongoose.Schema({
  fixtureId: { type: Number, required: true },
  leagueId: { type: Number },
  leagueName: { type: String, default: '' },
  homeTeamId: { type: Number },
  awayTeamId: { type: Number },
  homeTeam: { type: String, required: true },
  awayTeam: { type: String, required: true },
  homeLogo: { type: String, default: '' },
  awayLogo: { type: String, default: '' },
  kickoff: { type: Date, required: true },
  status: { type: String, default: '' },
}, { _id: false });

/**
 * Real-world prize attached to a normal pool. Distinct from the
 * `prize` / `prizeLabel` cash-display fields — this carries the
 * structured payload the iOS/web client needs to render the prize
 * hero (image asset key) plus admin operational data
 * (USD value for telemetry, free-text delivery note for fulfillment).
 *
 * Only set on admin-curated pools. Presence triggers AMOE + Apple
 * Guideline 5.3 disclaimers in the detail view client-side.
 */
const realPrizeSchema = new mongoose.Schema({
  label: { type: String, default: '' },
  prizeUSD: { type: Number, default: 0 },
  // Bundled-asset key the iOS app maps to an image catalog entry,
  // e.g. "PrizeAmazonGift". Future flexibility: when prizes vary,
  // we'll add a `prizeImageURL` for remote-hosted images.
  imageKey: { type: String, default: '' },
  // Admin paper trail for delivery (e.g. "Sent Amazon code via email
  // 2026-05-08 to user X"). Free text, not parsed by code.
  deliveryNote: { type: String, default: '' },
}, { _id: false });

/**
 * One rung of a prize_ladder pool. "If your correct-pick count is in
 * [min, max], you win prizeMXN pesos." Ranges let the spec's
 * "1–7 aciertos → $0" collapse into one tier, and ranges are
 * deliberately NOT assumed monotonic — the "0 aciertos → $100"
 * consolation sits below the dead zone. See src/lib/prizeLadder.js
 * for the lookup + validation helpers.
 */
const prizeTierSchema = new mongoose.Schema({
  min: { type: Number, required: true },
  max: { type: Number, required: true },
  prizeMXN: { type: Number, required: true, default: 0 },
  // Optional admin-set display label; clients otherwise build their own
  // localized label from min/max + prizeMXN.
  label: { type: String, default: '' },
}, { _id: false });

const quinielaSchema = new mongoose.Schema({
  name: { type: String, required: true, trim: true },
  description: { type: String, default: '' },
  // Legacy cash-prize label used by admin-curated pools (e.g. "$5,000 MXN").
  // User-created MVP pools leave this empty — they use `prizeLabel` instead.
  prize: { type: String, default: '' },
  // Legacy entry-cost string (e.g. "$15"). User-created MVP pools are free ("0").
  cost: { type: String, default: '0' },
  currency: { type: String, default: 'MXN' },
  startDate: { type: Date, required: true },
  endDate: { type: Date },
  fixtures: { type: [fixtureSchema], default: [] },
  // Admin-controlled flag. When true the pool is pinned to the client's
  // QUICK PLAY hero / featured carousel. Only toggleable by admins
  // (field-level guard in controller), not by the creator.
  featured: { type: Boolean, default: false, index: true },
  // User who created the pool. Legacy admin pools may leave this null.
  // Points at User (previously pointed at a non-existent 'Admin' model).
  createdBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', index: true },
  // 'public' → discoverable in Home. 'private' → only reachable via inviteCode / link.
  // Admin-created pools default to 'public'; user-created default to 'private'.
  visibility: { type: String, enum: ['public', 'private'], default: 'public', index: true },
  // Short alphanum code used in share links (e.g. futpools://p/ABC23456).
  // 8 chars, uppercase, no confusables (0/O/1/I). Generated at create time.
  inviteCode: { type: String, unique: true, sparse: true, index: true },
  // Free-text label for MVP user-pools ("the loser buys pizza"). Not money.
  // Legacy `prize` field is still used for admin-curated pools with cash prizes.
  prizeLabel: { type: String, default: '' },

  // ── Pool format (competition type) ────────────────────────────────────
  // Orthogonal to `fundingModel` (which describes WHO funds the prize).
  // `poolType` describes the SCORING/PAYOUT format:
  //   standard     — one winner takes the pot (existing behavior).
  //   prize_ladder — every player wins a fixed prize based on their
  //                  number of correct picks ("aciertos"), per `prizeLadder`.
  //                  Platform-funded: prizes are paid regardless of pot size.
  // Absent/empty on legacy pools → treated as 'standard'.
  poolType: {
    type: String,
    enum: ['standard', 'prize_ladder'],
    default: 'standard',
    index: true,
  },
  // The tiered payout table for prize_ladder pools (empty for standard).
  // Stored desc by `min` for top-to-bottom thermometer display.
  prizeLadder: { type: [prizeTierSchema], default: [] },

  // ── Economy / prize pool (Phase 2) ────────────────────────────────────
  // How the prize is funded:
  //   none       — free social pool, no prize mechanics (default)
  //   peer       — players stake coins; winner gets pot × (1 - rakePercent)
  //   platform   — FutPools funds the prize, locked only if min participants met
  //   sponsored  — user-funded prize (v3): creator pays prize×1.1, friends play free
  //   brand      — brand-funded prize (Fase 3)
  fundingModel: {
    type: String,
    enum: ['none', 'peer', 'platform', 'sponsored', 'brand'],
    default: 'none',
    index: true,
  },
  // Coins the prize pool holds. Funded by platform (admin Platform Event) or
  // by the creator (v3 Sponsored pools). Released to winner only if
  // entries.count >= minParticipants at first fixture kickoff.
  platformPrizeCoins: { type: Number, default: 0 },
  // Who paid for the prize. For Sponsored → the creator. For Platform Event →
  // the admin. Settlement uses this to route refunds when a pool dissolves.
  prizeFunderUserId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', index: true },
  // Floor for the prize to unlock. Below this, prize dissolves and any coin
  // entries are refunded.
  minParticipants: { type: Number, default: 1 },
  // Virtual-currency entry cost for 'peer' pools. 0 for free/platform pools.
  entryCostCoins: { type: Number, default: 0 },
  // House cut on settlement. simple_version splits 65/35: winner gets
  // 65% of the gross pot, the remaining 35% absorbs Stripe processing
  // (~3.6% + $3 MXN) and the FutPools platform fee. Default raised
  // from the legacy 10% so newly-created pools settle on the new
  // economics without a per-pool override.
  rakePercent: { type: Number, default: 35 },
  // Lifecycle of the pledged prize. See plan file for transitions.
  prizeLockStatus: {
    type: String,
    enum: ['pledged', 'locked', 'dissolved', 'paid'],
    default: 'pledged',
    index: true,
  },
  prizeUnlockedAt: { type: Date },
  // Settlement tracking for the cron.
  settlementStatus: {
    type: String,
    enum: ['pending', 'settled', 'refunded', 'cancelled'],
    default: 'pending',
    index: true,
  },
  settledAt: { type: Date },
  winnerUserIds: { type: [mongoose.Schema.Types.ObjectId], default: [] },
  // Set once the "last call" reminder has gone out (a few hours before
  // startDate, to opted-in non-entrants). One-shot per pool, see
  // lifecycleEmailService.remindClosingPools.
  closingReminderSentAt: { type: Date, default: null },

  // Admin-only. When set, the pool is rendered as a "real prize pool"
  // with the bundled asset image, AMOE + Apple disclaimers in the
  // detail view, and surfaced in the Home WEEKLY POOL · REAL PRIZE
  // teaser. Listed via GET /quinielas?realPrize=1.
  realPrize: { type: realPrizeSchema, default: null },

  // ── simple_version (Phase 1) ──────────────────────────────────────────
  // Stripe Checkout entry fee in **MXN pesos** (NOT cents — converted to
  // centavos at the Stripe API call site by ×100). Default 50 = $50 MXN
  // per the simple_version product spec. Stored even on master pools
  // (unused there) so a single QuinielaEntry schema can carry stripe
  // metadata regardless of branch. Admin can override per pool.
  entryFeeMXN: { type: Number, default: 50 },
  // Set when the admin marks a settled pool as paid out to the winner
  // (manual SPEI / bank transfer happens off-band). The presence of
  // this date is what removes the pool from the AdminPayouts list.
  winnerPaidAt: { type: Date, default: null },
  // Free-text record of the payout (e.g. "SPEI ref ABC123, 2026-05-15").
  // Not parsed by code — operational paper trail only.
  winnerPaidNote: { type: String, default: null },
  // Set when admin cancels the pool via /admin/pools/:id/cancel. Refunds
  // are dispatched per-entry; this field marks the pool as no longer
  // active. cancelledReason carries the admin's free-text justification.
  cancelledAt: { type: Date, default: null },
  cancelledReason: { type: String, default: null },

  // Set when the "pool locks in 10 min — review pending payments" Telegram
  // alert has been sent (see poolLockAlertService). One-shot per pool so the
  // 1-minute scheduler doesn't re-alert every tick during the lock window.
  lockAlertSentAt: { type: Date, default: null },

  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now },
});

module.exports = mongoose.model('Quiniela', quinielaSchema);
